### PR TITLE
chore: add storage provider

### PR DESCRIPTION
**Pull request by rodrigopavezi**
## Problem and Motivation
We talked about this PR in the daily today. This PR has a bunch of fixes for issues @aimensahnoun  was having when integrating Lit protocol provider with HQ PoC. I also implemented better handling of encryption and decryption so clients can choose whether decrypt requests or not.

## Summary by CodeRabbit

- **New Features**
	- Enhanced client initialization to support asynchronous operations and error handling.
	- Added methods to check encryption and decryption availability across multiple classes.
	- Introduced new dependencies for improved encryption functionalities.
	- Added a build script to ensure proper package preparation before publishing.
	- New properties and methods in various classes to manage encryption and decryption states.

- **Bug Fixes**
	- Improved error handling for transaction parsing related to decryption availability.
	- Added console logging for better debugging of session signature retrieval.
	- Refined error messages for clarity during transaction parsing.

- **Chores**
	- Updated dependencies for enhanced functionality and support.
	- Modified test configurations for better asynchronous handling and alignment with new imports.

### CODE REVIEW

File: packages/lit-protocol-cipher/package.json [hunk 1]
```diff
@@ -43,13 +43,16 @@
     "@lit-protocol/auth-helpers": "6.11.2",
     "@lit-protocol/constants": "6.11.2",
     "@lit-protocol/lit-node-client": "6.11.2",
+    "@lit-protocol/lit-node-client-nodejs": "^7.0.0",
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/lit-protocol-cipher/package.json]: I was changed

File: packages/epk-cipher/src/ethereum-private-key-cipher-provider.ts [hunk 2]
```diff
@@ -27,6 +27,24 @@ export default class EthereumPrivateKeyCipherProvider
...
+  }
+
+  /**
+   * Check if decryption is available
+   *
+   * @returns true if decryption is available
+   */
+  public isDecryptionAvailable(): boolean {
+    return true;
+  }
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/epk-cipher/src/ethereum-private-key-cipher-provider.ts:46]: I will fix that

File: packages/lit-protocol-cipher/src/lit-protocol-cipher-provider.ts [hunk 3]
```diff
@@ -236,32 +264,36 @@ export default class LitProvider implements CipherProviderTypes.ICipherProvider
...
       );
 
-      return await LitJsSdk.encryptString(
+      return await encryptString(
         {
           accessControlConditions: accessControlConditions,
           dataToEncrypt: stringifiedData,
         },
-        client,
+        this.client!,
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/lit-protocol-cipher/src/lit-protocol-cipher-provider.ts:279]: I will fix that

File: packages/lit-protocol-cipher/src/lit-protocol-cipher-provider.ts [hunk 4]
```diff
@@ -153,21 +171,22 @@ export default class LitProvider implements CipherProviderTypes.ICipherProvider
...
+          chain: this.chain,
+          capabilityAuthSigs: [capacityDelegationAuthSig],
+          resourceAbilityRequests: [
+            {
+              resource: litResource,
+              ability: LIT_ABILITY.AccessControlConditionDecryption,
+            },
+          ],
+          authNeededCallback,
+        })) || {};
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/lit-protocol-cipher/src/lit-protocol-cipher-provider.ts:185]: I will fix that

File: packages/lit-protocol-cipher/src/lit-protocol-cipher-provider.ts [hunk 5]
```diff
@@ -67,15 +72,33 @@ export default class LitProvider implements CipherProviderTypes.ICipherProvider
...
+        removeItem: (key: string) => localStorage.removeItem(key),
+        provider: localStorage,
+      };
+
+      this.client = new LitJsSdk.LitNodeClientNodeJs({
         litNetwork: this.network,
+        storageProvider,
       });
+
+      await this.client.connect();
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/lit-protocol-cipher/src/lit-protocol-cipher-provider.ts:101]: I will fix that

File: packages/epk-cipher/src/ethereum-private-key-cipher-provider.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/epk-cipher/src/ethereum-private-key-cipher-provider.ts:46]: I will fix that

File: packages/lit-protocol-cipher/src/lit-protocol-cipher-provider.ts [hunk 6]
```diff
@@ -220,6 +239,15 @@ export default class LitProvider implements CipherProviderTypes.ICipherProvider
...
   }
 
+  /**
+   * @function isEncryptionAvailable
+   * @description Checks if encryption is available.
+   * @returns {boolean} A boolean indicating if encryption is available.
+   */
+  public isEncryptionAvailable(): boolean {
+    return true;
+  }
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/lit-protocol-cipher/src/lit-protocol-cipher-provider.ts:249]: I will fix that

File: packages/lit-protocol-cipher/src/lit-protocol-cipher-provider.ts [hunk 7]
```diff
@@ -67,15 +72,33 @@ export default class LitProvider implements CipherProviderTypes.ICipherProvider
...
-      return new LitJsSdk.LitNodeClient({
+      this.client = new LitJsSdk.LitNodeClient({
         litNetwork: this.network,
       });
+      await this.client.connect();
     } else {
-      return new LitJsSdk.LitNodeClientNodeJs({
+      // Create dynamic import URL to prevent webpack analysis
+      const moduleUrl = 'node-localstorage';
+      const { LocalStorage } = await (0, eval)(`import('${moduleUrl}')`);
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/lit-protocol-cipher/src/lit-protocol-cipher-provider.ts:85]: That won't work on nextjs webpack

File: packages/lit-protocol-cipher/test/index.test.ts [hunk 8]
```diff
@@ -6,60 +6,55 @@ import { HttpDataAccess, NodeConnectionConfig } from '@requestnetwork/request-cl
...
     mockSigner = {
-      getAddress: jest.fn().mockReturnValueOnce(mockWalletAddress),
-      signMessage: jest.fn().mockReturnValueOnce('mock-signature'),
+      getAddress: jest.fn().mockReturnValue(Promise.resolve(mockWalletAddress)),
+      signMessage: jest.fn().mockReturnValue(Promise.resolve('mock-signature')),
     } as unknown as jest.Mocked<Signer>;
 
-    // Initialize LitProvider
     litProvider = new LitProvider(mockChain, mockNetwork, mockNodeConnectionConfig);
+    await new Promise((resolve) => setTimeout(resolve, 100));
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/lit-protocol-cipher/test/index.test.ts]: I fixed that

File: packages/lit-protocol-cipher/package.json [hunk 9]
```diff
@@ -35,21 +35,26 @@
...
-    "@lit-protocol/constants": "6.11.2",
-    "@lit-protocol/lit-node-client": "6.11.2",
-    "@lit-protocol/types": "6.11.2",
+    "@lit-protocol/auth-browser": "7.0.0",
+    "@lit-protocol/auth-helpers": "7.0.0",
+    "@lit-protocol/constants": "7.0.0",
+    "@lit-protocol/encryption": "7.0.0",
+    "@lit-protocol/lit-node-client": "7.0.0",
+    "@lit-protocol/lit-node-client-nodejs": "7.0.0",
+    "@lit-protocol/types": "7.0.0",
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/lit-protocol-cipher/package.json:49]: I will fix that

File: packages/lit-protocol-cipher/test/index.test.ts [hunk 10]
```diff
@@ -87,7 +82,7 @@ describe('LitProvider', () => {
 
       await litProvider.disconnectWallet();
 
-      expect(disconnectWeb3).not.toHaveBeenCalled();
+      // No specific assertion here since it just sets sessionSigs to null
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/lit-protocol-cipher/test/index.test.ts:86]: I will fix that

File: packages/lit-protocol-cipher/test/index.test.ts [hunk 11]
```diff
@@ -150,56 +149,7 @@ describe('LitProvider', () => {
...
-
-      try {
-        await litProvider.encrypt(mockData, {
-          encryptionParams: mockEncryptionParams,
-        });
-      } catch (error) {
-        expect(error.message).toBe('Encryption failed');
-      }
-    });
+    // ... (rest of your encrypt tests with necessary adjustments)
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/lit-protocol-cipher/test/index.test.ts:152]: This is fine for now

File: packages/request-node/src/config.ts [hunk 12]
```diff
@@ -47,7 +47,7 @@ const defaultValues = {
   wallet: {
     mnemonic: 'candy maple cake sugar pudding cream honey rich smooth crumble sweet treat',
   },
-  litProtocolNetwork: LitNetwork.DatilTest,
+  litProtocolNetwork: LIT_NETWORK.DatilTest,
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/request-node/src/config.ts:50]: I will leave like that for now

File: packages/lit-protocol-cipher/src/lit-protocol-cipher-provider.ts [hunk 13]
```diff
@@ -54,28 +70,58 @@ export default class LitProvider implements CipherProviderTypes.ICipherProvider
...
+          debug: this.debug,
+        });
+        await this.client.connect();
+      } else {
+        // Evaluate the code in a way that prevents static analysis
+        const getNodeStorage = new Function(
+          `
+          return import('node-localstorage').then(m => m.LocalStorage);
+          `,
+        );
```
[the remark this answers is not in the document]
Comment (rodrigopavezi) [reply packages/lit-protocol-cipher/src/lit-protocol-cipher-provider.ts:102]: I will skip this for now

**Review by aimensahnoun — APPROVED** [review]

**Review by sstefdev — APPROVED** [review]

**Comment by MantisClone** [comment]
After this was merged, @rodrigopavezi and I discussed alternative designs for the ICipherProvider interface and decided that this was the least bad option.

Alternatives included:
1. `skipDecryption` argument when calling `fromIdentity()`, `fromTopic()`, or `fromRequestId()`. - This might be easier to use, but we don't want to propagate a `skipDecryption` up through all the other software layers. 
2. Immutable `ICipherProvider` where `decryptionEnabled` is not mutatable after construction. - This is more difficult to use and wouldn't provide any benefit.

**Pull request merged**